implementations/rust/ockam/ockam_core/src/error/inner/mod.rs

```diff
@@ -0,0 +1,183 @@
+#[allow(unused)]
+use super::code::{Kind, Origin};
+#[cfg(feature = "std")]
+use crate::compat::{borrow::ToOwned, vec::Vec};
+use crate::compat::{boxed::Box, error::Error as ErrorTrait, string::String};
+use crate::error::code::ErrorCode;
+use serde::{Deserialize, Serialize};
+
+mod formatting;
+#[cfg(all(feature = "std", feature = "error-traces"))]
+mod trace_config;
+
+type BoxDynErr = Box<dyn ErrorTrait + Send + Sync + 'static>;
+
+/// The internal error type, which [`super::NewError`] wraps.
+///
+/// When an allocator is available, this is `Box`ed for performance reasons,
+/// (although some of the fields here are a little goofy in low level contexts).
+#[derive(Serialize, Deserialize)]
+pub(super) struct ErrorData {
```

Comment:
It looks like for hypothetical no_std vault developer the only usable place to report information about the error is `extra: i32` field, which is probably worse than it was.
std situation is much improved though

Reply:
> One thing it might be quite useful for is to stick a 32-bit memory address in.

It's tricky because this part probably should be serializable. On embedded we could store a `Box<Box<dyn core2::error::Error>>` in the non-error code portion (or [`ThinBox<core2::error::Error>`](https://doc.rust-lang.org/nightly/std/boxed/struct.ThinBox.html)). Or a `NonNull<()>` that you take responsibility for wrangling (it would be leaked though, and wouldn't track lifetimes, so this would be quite unsafe).

Reply:
`NonNull<()>` sounds about the closest to what I have in mind. The two different use-cases I thought of are:

1. pass a rusty pointer to [`defmt`](https://defmt.ferrous-systems.com/)  
2. pass a raw memory-address to a hardware debugger

What's the most rust-idiomatic solution we can come up with that fits both?

That said, this is not a particularly short-term design decision we need to make, just thinking about keeping escape hatches open that might be useful down the road?